src/utils/OpenAIStream.ts

```diff
@@ -0,0 +1,80 @@
+import {
+    createParser,
+    ParsedEvent,
+    ReconnectInterval,
+  } from 'eventsource-parser';
+  
+  export type ChatGPTAgent = 'user' | 'system';
+  
+  export interface ChatGPTMessage {
+    role: ChatGPTAgent;
+    content: string;
+  }
+  
+  export interface OpenAIStreamPayload {
+    model: string;
+    messages: ChatGPTMessage[];
+    temperature: number;
+    top_p: number;
+    frequency_penalty: number;
+    presence_penalty: number;
+    max_tokens: number;
+    stream: boolean;
+    n: number;
+  }
+  
+  export async function OpenAIStream(payload: OpenAIStreamPayload) {
+    const encoder = new TextEncoder();
+    const decoder = new TextDecoder();
+  
+    let counter = 0;
+  
+    const res = await fetch('https://api.openai.com/v1/chat/completions', {
+      headers: {
+        'Content-Type': 'application/json',
+        Authorization: `Bearer ${process.env.OPENAI_API_KEY ?? ''}`,
+      },
+      method: 'POST',
+      body: JSON.stringify(payload),
+    });
+  
+    const stream = new ReadableStream({
+      async start(controller) {
+        // callback
+        function onParse(event: ParsedEvent | ReconnectInterval) {
+          if (event.type === 'event') {
+            const data = event.data;
+            // https://beta.openai.com/docs/api-reference/completions/create#completions/create-stream
+            if (data === '[DONE]') {
+              controller.close();
+              return;
+            }
+            try {
+              const json = JSON.parse(data);
+              const text = json.choices[0].delta?.content || '';
+              if (counter < 2 && (text.match(/\n/) || []).length) {
+                // this is a prefix character (i.e., "\n\n"), do nothing
+                return;
+              }
+              const queue = encoder.encode(text);
+              controller.enqueue(queue);
+              counter++;
+            } catch (e) {
+              // maybe parse error
+              controller.error(e);
+            }
+          }
+        }
+  
+        // stream response (SSE) from OpenAI may be fragmented into multiple chunks
+        // this ensures we properly read chunks and invoke an event for each SSE event stream
+        const parser = createParser(onParse);
+        // https://web.dev/streams/#asynchronous-iteration
+        for await (const chunk of res.body as any) {
+          parser.feed(decoder.decode(chunk));
+        }
+      },
+    });
+  
+    return stream;
+  }
```

Comment:
@youssefadr [nextjs noob here] il faudrait pas mettre ce code dans le backend pour ne pas exposer la clé API?

Reply:
@louis-sanna-perso La clé API sera dans un fichier .env au niveau de la racine. J'ai oublié de le rajouter mais il contiendrait la clé comme ceci OPENAI_API_KEY= , à moins qu'elle soit exposée autrement avec ce code.

Pour le code que je propose ici, je l'ai trouvé sur ce auto https://www.youtube.com/watch?v=JcE-1xzQTE0, si tu veux regarder les détails.

Reply:
@youssefadr en react SPA les variables d'env front sont accessible par tous. Je sais pas comme c'est pour Nextjs, à checker.

Reply:
Ah d'accord je vois ! 
[AConfirmer] J'ai regardé la doc https://nextjs.org/docs/basic-features/environment-variables, et ils ont l'air de dire que pour exposer les variables au browser il faut rajouter un préfixe NEXT_PUBLIC, ça passe du coup ?

Reply:
@youssefadr nickel!

you peux rajouter la partie front pour checker que ça marche?

(ou peut mettre la prompte en dur, je veux juste voir la stream de la réponse s'afficher)

Reply:
Je pense que @louis-sanna-perso a raison et que la fonction utils doit être importé et utilisée dans la partie serveur, puis que le  code dans pages y fasse référence. c'est pas trop différent d'une query de db sur le principe je pense

Reply:
@youssefadr @youssefadr alors après discussion avec ChatGPT, `pages/api` ça à l'air d'être dans le back, et là on run le code dans une edge function pour streamer la réponse le plus rapidement possible. Donc le code @youssefadr est correct!

<img width="971" alt="Screenshot 2023-03-10 at 20 49 09" src="https://user-images.githubusercontent.com/4759058/224414088-ae774558-f257-48d0-8ec2-797c37d5c7ab.png">
<img width="1077" alt="Screenshot 2023-03-10 at 20 52 29" src="https://user-images.githubusercontent.com/4759058/224414794-ce05bfc8-4d8e-413f-906d-45ced91b0aa5.png">
<img width="1086" alt="Screenshot 2023-03-10 at 20 52 40" src="https://user-images.githubusercontent.com/4759058/224414798-ef19e04a-b639-4d4e-bd4a-15ef83db005d.png">